### PR TITLE
Cleanup of enchantments

### PR DESCRIPTION
**Pull request by parlough**
**SpongeAPI** | [SpongeCommon](https://github.com/SpongePowered/SpongeCommon/pull/1593) | [SpongeVanilla](https://github.com/SpongePowered/SpongeVanilla/pull/361) | [SpongeForge](https://github.com/SpongePowered/SpongeForge/pull/1847)

I will finish what I need to tomorrow, but for now looking for any feedback and also where I undoubtedly made mistakes as I haven't really tested much and some stuff was done automatically through IntelliJ refactoring and I may have missed mistakes.

Closes https://github.com/SpongePowered/SpongeAPI/issues/1684

### CODE REVIEW

File: src/main/java/org/spongepowered/api/data/Queries.java [hunk 1]
```diff
@@ -42,8 +42,8 @@
     public static final DataQuery WEIGHTED_SERIALIZABLE = of("DataSerializable");
     public static final DataQuery WEIGHTED_SERIALIZABLE_WEIGHT = of("Weight");
 
-    // Item Enchantment
-    public static final DataQuery ENCHANTMENT_ID = of("Enchantment");
+    // Item EnchantmentType
+    public static final DataQuery ENCHANTMENT_ID = of("EnchantmentType");
```
Comment (parlough) [inline src/main/java/org/spongepowered/api/data/Queries.java]: This was done automatically by IntelliJ. Should I keep I revert this change to not cause compatibility issues?

File: src/main/java/org/spongepowered/api/block/tileentity/EnchantmentTable.java [hunk 2]
```diff
@@ -25,7 +25,7 @@
 package org.spongepowered.api.block.tileentity;
 
 /**
- * Represents an Enchantment Table.
+ * Represents an EnchantmentType Table.
```
Comment (liach) [inline src/main/java/org/spongepowered/api/block/tileentity/EnchantmentTable.java]: This one is absolutely wrong.

File: src/main/java/org/spongepowered/api/data/manipulator/catalog/CatalogItemData.java [hunk 3]
```diff
@@ -114,7 +114,7 @@
      */
     public static final Class<DyeableData> DYEABLE_DATA = DyeableData.class;
     /**
-     * Signifies that an item has {@link Enchantment}s applied to it.
+     * Signifies that an item has {@link EnchantmentType}s applied to it.
```
Comment (liach) [inline src/main/java/org/spongepowered/api/data/manipulator/catalog/CatalogItemData.java]: You probably should revert this one.

File: src/main/java/org/spongepowered/api/data/manipulator/catalog/CatalogItemData.java [hunk 4]
```diff
@@ -169,7 +169,7 @@
      */
     public static final Class<SpawnableData> SPAWNABLE_DATA = SpawnableData.class;
     /**
-     * Signifies that there are stored {@link Enchantment}s available on the
+     * Signifies that there are stored {@link EnchantmentType}s available on the
```
Comment (liach) [inline src/main/java/org/spongepowered/api/data/manipulator/catalog/CatalogItemData.java]: Same?

File: src/main/java/org/spongepowered/api/data/meta/Enchantment.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.data.meta;
```
Comment (liach) [inline src/main/java/org/spongepowered/api/data/meta/Enchantment.java]: I think we should move the package of this class. It is not really a meta any more, since it has been converted into an interface.

File: src/main/java/org/spongepowered/api/data/meta/Enchantment.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+import org.spongepowered.api.data.persistence.DataBuilder;
+import org.spongepowered.api.item.EnchantmentType;
+import org.spongepowered.api.item.inventory.ItemStack;
+
+/**
+ * Represents an {@link EnchantmentType} on an {@link ItemStack} that is paired
+ * with a level.
+ *
+ * <p>The contract of enchantments is that their level will
+ * never be below 1.</p>
```
Comment (liach) [inline src/main/java/org/spongepowered/api/data/meta/Enchantment.java]:
````
<p>The contract of enchantments is that their level will
be in the range of {@code getType().getMinimumLevel()}
and {@code getType().getMinimumLevel()}.</p>
````

File: src/main/java/org/spongepowered/api/item/EnchantmentTypes.java [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.item;
+
+import org.spongepowered.api.util.generator.dummy.DummyObjectProvider;
+
+/**
+ * An enumeration of known {@link EnchantmentType}s.
```
Comment (liach) [inline src/main/java/org/spongepowered/api/item/EnchantmentTypes.java]: You should probably mention that all of them have a minimum level of 1.

File: src/main/java/org/spongepowered/api/item/EnchantmentTypes.java [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+    /**
+     * Increases damages to arthropod mobs. In vanilla this includes spiders,
+     * cave spiders, silverfish, and endermites.
+     *
+     * <p>In vanilla the maximum level is 5.</p>
+     */
+    public static final EnchantmentType BANE_OF_ARTHROPODS = DummyObjectProvider.createFor(EnchantmentType.class, "BANE_OF_ARTHROPODS");
+
+    /**
+     * Prevents removal of the cursed items.
```
Comment (liach) [inline src/main/java/org/spongepowered/api/item/EnchantmentTypes.java]: You should probably state that this enchantment only target items in the 4 armor slots.

File: src/main/java/org/spongepowered/api/item/EnchantmentTypes.java [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+     *
+     * <p>In vanilla the maximum level is 1.</p>
+     */
+    public static final EnchantmentType AQUA_AFFINITY = DummyObjectProvider.createFor(EnchantmentType.class, "AQUA_AFFINITY");
+
+    /**
+     * Increases damages to arthropod mobs. In vanilla this includes spiders,
+     * cave spiders, silverfish, and endermites.
+     *
+     * <p>In vanilla the maximum level is 5.</p>
```
Comment (liach) [inline src/main/java/org/spongepowered/api/item/EnchantmentTypes.java]: At high level, the enchantment will slow down the arthropod hit for few seconds.

File: src/main/java/org/spongepowered/api/item/EnchantmentTypes.java [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+
+/**
+ * An enumeration of known {@link EnchantmentType}s.
+ */
+public final class EnchantmentTypes {
+
+    // SORTFIELDS:ON
+
+    /**
+     * Increases underwater mining rate.
```
Comment (liach) [inline src/main/java/org/spongepowered/api/item/EnchantmentTypes.java]: You should say it just increase the speed to the regular mining speed.

File: src/main/java/org/spongepowered/api/data/meta/Enchantment.java (hunk 6 above)
Comment (liach) [reply src/main/java/org/spongepowered/api/data/meta/Enchantment.java]: Uhh yes, it can be any natural number as long as the number can be stored in data.
Does this hurt items with empty enchantment tags? Those tags may give item stacks an effect while not having a valid enchantment.

File: src/main/java/org/spongepowered/api/item/EnchantmentTypes.java [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+     * Reduces projectile damage you take, for example from arrows, ghasts,
+     * blaze fire charges, and similar in vanilla.
+     *
+     * <p>In vanilla the maximum level is 4.</p>
+     */
+    public static final EnchantmentType PROJECTILE_PROTECTION = DummyObjectProvider.createFor(EnchantmentType.class, "PROJECTILE_PROTECTION");
+
+    /**
+     * Reduces all damage, outside of a few sources, such as the void,
+     * the kill command, and hunger damage in vanilla.
```
Comment (liach) [inline src/main/java/org/spongepowered/api/item/EnchantmentTypes.java]: Probably mention that it doesn't reduce damage that bypasses armor.

File: src/main/java/org/spongepowered/api/item/EnchantmentTypes.java [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+
+    /**
+     * Attackers are damaged when they deal damage to the wearer.
+     *
+     * <p>In vanilla the maximum level is 3.</p>
+     */
+    public static final EnchantmentType THORNS = DummyObjectProvider.createFor(EnchantmentType.class, "THORNS");
+
+    /**
+     * RIncreases effective durability.
```
Comment (liach) [inline src/main/java/org/spongepowered/api/item/EnchantmentTypes.java]: Remove that R I guess

File: src/main/java/org/spongepowered/api/util/TypeTokens.java [hunk 13]
```diff
@@ -213,9 +213,9 @@
 
     public static final TypeToken<List<FireworkEffect>> LIST_FIREWORK_TOKEN = new TypeToken<List<FireworkEffect>>() {private static final long serialVersionUID = -1;};
 
-    public static final TypeToken<List<ItemEnchantment>> LIST_ITEM_ENCHANTMENT_TOKEN = new TypeToken<List<ItemEnchantment>>() {private static final long serialVersionUID = -1;};
+    public static final TypeToken<List<Enchantment>> LIST_ITEM_ENCHANTMENT_TOKEN = new TypeToken<List<Enchantment>>() {private static final long serialVersionUID = -1;};
 
-    public static final TypeToken<ListValue<ItemEnchantment>> LIST_ITEM_ENCHANTMENT_VALUE_TOKEN = new TypeToken<ListValue<ItemEnchantment>>() {private static final long serialVersionUID = -1;};
+    public static final TypeToken<ListValue<Enchantment>> LIST_ITEM_ENCHANTMENT_VALUE_TOKEN = new TypeToken<ListValue<Enchantment>>() {private static final long serialVersionUID = -1;};
```
Comment (liach) [inline src/main/java/org/spongepowered/api/util/TypeTokens.java]: You should rename these 2 fields.

File: src/main/java/org/spongepowered/api/data/meta/Enchantment.java (hunk 6 above)
Comment (parlough) [reply src/main/java/org/spongepowered/api/data/meta/Enchantment.java]: So we only check when building if it is above 0 or below Short.MAX_VALUE. There are issues if the level is below 0 if I recall correctly but in reality it can be 0(I think?). I just changed it to 1 since I didn't see a reason why an enchantment would ever need to be 0 and Minecraft doesn't really expect it with all of theirs having a minimum level of 0. Do you think I should switch back to 0 or greater?

Also to your suggestion. I don't think that is a great idea as we return the vanilla max and min which are not necessarily as high as they could be set.

I'm heavily open to feedback relating to our lower limit and how to specify it.

File: src/main/java/org/spongepowered/api/item/EnchantmentTypes.java (hunk 9 above)
Comment (parlough) [reply src/main/java/org/spongepowered/api/item/EnchantmentTypes.java]: Interesting, never knew this.

File: src/main/java/org/spongepowered/api/item/EnchantmentTypes.java (hunk 8 above)
Comment (parlough) [reply src/main/java/org/spongepowered/api/item/EnchantmentTypes.java]: Makes sens, will do.

File: src/main/java/org/spongepowered/api/data/Queries.java (hunk 1 above)
Comment (parlough) [reply src/main/java/org/spongepowered/api/data/Queries.java]: Anyone have input here?

Comment (kashike) [reply src/main/java/org/spongepowered/api/data/Queries.java]: undo

File: src/main/java/org/spongepowered/api/data/meta/Enchantment.java (hunk 5 above)
Comment (parlough) [reply src/main/java/org/spongepowered/api/data/meta/Enchantment.java]: How do you feel about moving `Enchantment`, `EnchantmentType`, and `EnchantmentTypes` to a new `enchantment` package inside of the `effect` package?

File: src/main/java/org/spongepowered/api/item/EnchantmentTypes.java (hunk 7 above)
Comment (parlough) [reply src/main/java/org/spongepowered/api/item/EnchantmentTypes.java]: Done.

File: src/main/java/org/spongepowered/api/data/Queries.java (hunk 1 above)
Comment (parlough) [reply src/main/java/org/spongepowered/api/data/Queries.java]: Will do. 👍

**Comment by Cybermaxke** [comment]
Could we also use a custom ListValue and extra methods in the enchantment data manipulator to turn in some kind of mapped data? So you don't have to loop through all the enchantments to get the level, this could also be optimized internally.

```java
import org.spongepowered.api.data.value.mutable.ListValue;
import org.spongepowered.api.item.EnchantmentType;
import org.spongepowered.api.item.Enchantment;

import java.util.OptionalInt;

public interface EnchantmentListValue extends ListValue<Enchantment> {

    OptionalInt getLevel(EnchantmentType enchantment);

    void add(EnchantmentType enchantment, int level);
}
```

**Comment by ryantheleach** [comment]
No we can not @Cybermaxke :( 

Vanilla very occasionally will generate with multiple enchants of the same type on an item.

https://www.youtube.com/watch?v=66zHMfi-bsI

**Comment by parlough** [comment]
What are people's opinions on the minimum enchantment value? Should we allow it to be 0 or above, or 1 and above to match vanilla's enchantments and to avoid possible issues.

**Comment by liach** [comment]
@Meronat We should probably allow only positive levels, no zero. The glowing effects some items have is from an empty enchantment tag list (so the list can exist and be empty, a different concept) so level 0 enchantments are useless

**Comment by parlough** [comment]
These 4 PRs should be ready for more testing and review. 🥂

**Comment by ryantheleach** [comment]
@Meronat are we exposing the 'natural maximum' enchantment level of each item anywhere? Is it even plausible?

**Comment by parlough** [comment]
@ryantheleach I believe this is what you mean? https://github.com/Meronat/SpongeAPI/blob/ae6f6d92a545bdd9ff8fd40b08f77c5b724b87f8/src/main/java/org/spongepowered/api/item/enchantment/EnchantmentType.java#L62

**Comment by ryantheleach** [comment]
https://gaming.stackexchange.com/questions/262999/what-is-the-highest-effective-enchantment-level/263013#263013 States that some out of band negative values may have some effect. Not sure if we should handle them, or reject them outright. How do other parts of the API behave with out of band values?

Also don't know if any mods do anything outrageous.

File: src/main/java/org/spongepowered/api/item/enchantment/Enchantment.java [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+         * @return The modified builder, for chaining
+         */
+        Builder type(EnchantmentType enchantmentType);
+
+        /**
+         * Sets the level for this enchantment.
+         *
+         * @param level The desired level
+         * @return The modified builder for chaining
+         * @throws IllegalArgumentException If the level is below 1
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/item/enchantment/Enchantment.java]: This contradicts the argument in the java doc for the class of `Enchantment` [here](https://github.com/SpongePowered/SpongeAPI/pull/1687/files#diff-5e7629a1c106a95db1ab90b7621fc617R37)

Comment (gabizou) [reply src/main/java/org/spongepowered/api/item/enchantment/Enchantment.java]: And I'm somewhat concerned about mod compatibility here.

**Review by gabizou — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/api/item/enchantment/Enchantment.java (hunk 14 above)
Comment (parlough) [reply src/main/java/org/spongepowered/api/item/enchantment/Enchantment.java]: There I said their level will always be larger than 0, which equates to the same thing as never below 1, but I can clean up the wording and add the max as well. 

How do you propose I handle the concern? Should we simply allow them to use the full range of `Short.MIN_VALUE` to `Short.MAX_VALUE`, specifying that unintentional issues may arise if you go outside of `EnchantmentType#getMinumumLevel` and `EnchantmentType#getMaxmimumLevel`?

Comment (gabizou) [reply src/main/java/org/spongepowered/api/item/enchantment/Enchantment.java]: > Should we simply allow them to use the full range of Short.MIN_VALUE to Short.MAX_VALUE, specifying that unintentional issues may arise if you go outside of EnchantmentType#getMinumumLevel and EnchantmentType#getMaxmimumLevel?

I am going to have to say "yes" because we cannot 100% control the possibilities of enchantments. If developers choose to make unworkable enchantments, that's more on them.

Comment (parlough) [reply src/main/java/org/spongepowered/api/item/enchantment/Enchantment.java]: I will make this change and the necessary changes to the docs later tomorrow.

**Edit:** Sorry haven't had a chance to get around to this yet, will do soon!

**Comment by parlough** [comment]
Any final reviews on API or implementation? Changes work for me.

**Review by gabizou — APPROVED** [review]
